File: packages/core/src/profiles/service.ts
import type { Session } from '../auth'
import type { Paginated } from '../common'
import type { ValueGenerator } from '../utils'
import { ProfileNotFoundError } from './error'
import type { ProfilesRepository } from './repository'
import type { Profile, ProfileDto } from './types'

export class ProfilesService {
	constructor(
		private repo: ProfilesRepository,
		private idGenerator: ValueGenerator<string>,
	) {}

	async list(session: Session): Promise<Paginated<Profile>> {
		const profiles = await this.repo.findMany({
			userId: session.userId,
		})

		return {
			hasMore: false,
			data: profiles,
		}
	}

	async get(id: string, session: Session): Promise<Profile> {
		const profile = await this.repo.findOne({
			id,
			userId: session.userId,
		})
		if (!profile) {
			throw new ProfileNotFoundError(id)
		}

		return profile
	}

	async create(dto: ProfileDto, session: Session): Promise<Profile> {
		const profile = await this.repo.create({
			id: this.idGenerator.generate(),
			title: dto.title,
			imageUrl: dto.imageUrl,
			isDefault: false,
			userId: session.userId,
		})

		return profile
	}

	async delete(id: string, session: Session): Promise<void> {
		const deleted = await this.repo.delete({
			id,
			userId: session.userId,
		})
		if (!deleted) {
			throw new ProfileNotFoundError(id)
		}
	}
}

File: crates/core/src/profiles/service.rs
use std::sync::Arc;

use super::{
    CreateProfile, Error, ProfileCreateData, ProfileFindByIdParams, ProfileFindManyParams,
    ProfileFindOneParams, ProfilesRepository, UpdateProfile,
};
use crate::{
    common::{Paginated, Session},
    Profile,
};

pub struct ProfilesService {
    repo: Arc<dyn ProfilesRepository + Send + Sync>,
}

impl ProfilesService {
    pub fn new(repo: Arc<dyn ProfilesRepository + Send + Sync>) -> Self {
        Self { repo }
    }

    pub async fn list(&self, session: Session) -> Result<Paginated<Profile>, Error> {
        let params = ProfileFindManyParams {
            user_id: session.user_id,
        };
        let profiles = self.repo.find_many(params).await?;

        let paginated = Paginated::<Profile> {
            has_more: false,
            data: profiles,
        };

        Ok(paginated)
    }

    pub async fn get(&self, id: String, session: Session) -> Result<Profile, Error> {
        let params = ProfileFindByIdParams {
            id,
            user_id: session.user_id,
        };
        let params = ProfileFindOneParams::ById(params);
        let profile = self.repo.find_one(params).await?;

        Ok(profile)
    }

    pub async fn create(&self, data: CreateProfile, session: Session) -> Result<Profile, Error> {
        let data = ProfileCreateData {
            title: data.title,
            image_url: data.image_url,
            user_id: session.user_id,
        };
        let profile = self.repo.create(data).await?;

        Ok(profile)
    }

    pub async fn update(
        &self,
        id: String,
        data: UpdateProfile,
        session: Session,
    ) -> Result<Profile, Error> {
        let params = ProfileFindByIdParams {
            id,
            user_id: session.user_id,
        };
        let profile = self.repo.update(params, data.into()).await?;

        Ok(profile)
    }

    pub async fn delete(&self, id: String, session: Session) -> Result<(), Error> {
        let params = ProfileFindByIdParams {
            id,
            user_id: session.user_id,
        };
        self.repo.delete(params).await?;

        Ok(())
    }
}
